### PR TITLE
Rate Limit headers

### PR DESCRIPTION
**Pull request by eloycoto**
This PR adds a policy that enables rate-limits headers per service. 

All changes explained in the commit messages.

### CODE REVIEW

**Comment by unleashed** [comment]
FYI @ioggstream

**Comment by eloycoto** [comment]
@davidor out for review

**Comment by davidor** [comment]
@unleashed as you are one of the authors of the RFC, it would be good if you could take a look at this PR. In particular, it would be great if you could review the documentation and the examples in the integration tests.

File: gateway/src/apicast/policy/apicast/apicast.lua [hunk 1]
```diff
@@ -102,7 +103,7 @@ function _M:access(context)
   local p = context and context.proxy or ctx.proxy or self.proxy
 
   if p then
-    return p:access(context.service, context.usage, context.credentials, context.ttl)
+    return p:access(context, context.service, context.usage, context.credentials, context.ttl)
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply gateway/src/apicast/policy/apicast/apicast.lua]: no, it's getting called from batcher policy and all it's needed.

File: gateway/src/apicast/policy/rate_limit_headers/cache.lua [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  if err then
+      ngx.log(ngx.ERR, "Cannot start cache for usage metrics, err=", err)
+      return err
+  end
+  self.cache = cache
+  self.namespace = namespace
+  return self
+end
+
+function _M:get_key(usage)
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply gateway/src/apicast/policy/rate_limit_headers/cache.lua]: It's to be able to test easily and make code more readable.

File: gateway/src/apicast/policy/rate_limit_headers/rate_limit_headers.lua [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+local usage_cache = require "cache"
+
+local new = _M.new
+local get_phase = ngx.get_phase
+
+local threescale_limit_header = "3scale-Limit-Max-Value"
+local threescale_remaining_header = "3scale-Limit-Remaining"
+local threescale_reset_header = "3scale-Limit-Reset"
+
+local limit_header = "RateLimit-Limit"
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply gateway/src/apicast/policy/rate_limit_headers/rate_limit_headers.lua]: Done.

File: t/apicast-policy-rate_limit_headers.t [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+--- request eval
+["GET /foo?user_key=foo", "GET /foo?user_key=foo"]
+--- response_headers eval
+--- response_headers_like eval
+[
+"RateLimit-Limit: 10\n" .
+    "RateLimit-Remaining: 9\n".
+    "RateLimit-Reset: [56]\\d\n",
+"RateLimit-Limit: 10\n" .
+    "RateLimit-Remaining: 7\n".
```
[the remark this answers is not in the document]
Comment (eloycoto) [reply t/apicast-policy-rate_limit_headers.t]: Ok, had a chat with Alex and he also agrees with. Thanks!

[the remark this answers is not in the document]
Comment (unleashed) [reply t/apicast-policy-rate_limit_headers.t]: :+1:

**Pull request merged**